File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/NomisEventOutcomeMapper.java
package uk.gov.justice.digital.hmpps.whereabouts.services;

import uk.gov.justice.digital.hmpps.whereabouts.model.AbsentReason;

import java.util.HashMap;

class NomisEventOutcomeMapper {
    private static final HashMap<AbsentReason, EventOutcome> eventOutcomes = new HashMap<>();

    NomisEventOutcomeMapper() {
        //paid absences
        eventOutcomes.put(AbsentReason.AcceptableAbsence, new EventOutcome("ACCAB", null, null));
        eventOutcomes.put(AbsentReason.NotRequired, new EventOutcome("NREQ", null, null));
        eventOutcomes.put(AbsentReason.ApprovedCourse, new EventOutcome("ACCAB", null, null));

        // unpaid absences
        eventOutcomes.put(AbsentReason.SessionCancelled, new EventOutcome("CANC", null, null));
        eventOutcomes.put(AbsentReason.RestInCell, new EventOutcome("REST", null, null));

        eventOutcomes.put(AbsentReason.Sick, new EventOutcome("REST", null, null));
        eventOutcomes.put(AbsentReason.RestDay, new EventOutcome("REST", null, null));

        //unpaid absences that trigger automatic iep warnings
        eventOutcomes.put(AbsentReason.Refused, new EventOutcome("UNACAB", null, null));
        eventOutcomes.put(AbsentReason.UnacceptableAbsence, new EventOutcome("UNACAB", null, null));

    }

    EventOutcome getEventOutcome(final AbsentReason reason, final boolean attended, final boolean paid, final String comment) {

        if (attended && reason != null) {
            throw new IllegalArgumentException("An absent reason was supplied for a valid attendance");
        }

        if (!attended && reason == null) {
            throw new IllegalArgumentException("An absent reason was not supplied");
        }

        if (attended && paid)
            return new EventOutcome("ATT", "STANDARD", comment);

        if (paid && !AbsentReason.getPaidReasons().contains(reason)) {
            throw new IllegalArgumentException(String.format("%s is not a valid paid reason", reason));
        }

        if (!paid && !AbsentReason.getUnpaidReasons().contains(reason)) {
            throw new IllegalArgumentException(String.format("%s is not a valid unpaid reason", reason));
        }

        var outcome = eventOutcomes.get(reason);

        if (comment != null) {
            outcome.setOutcomeComment(comment);
        }

        return outcome;

    }
}

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/NomisEventOutcomeMapper.kt
package uk.gov.justice.digital.hmpps.whereabouts.services

import org.springframework.stereotype.Service
import uk.gov.justice.digital.hmpps.whereabouts.model.AbsentReason
import uk.gov.justice.digital.hmpps.whereabouts.model.AbsentReason.Companion.paidReasons
import uk.gov.justice.digital.hmpps.whereabouts.model.AbsentReason.Companion.unpaidReasons
import uk.gov.justice.digital.hmpps.whereabouts.model.AbsentSubReason

@Service
class NomisEventOutcomeMapper {
  fun getEventOutcome(reason: AbsentReason?, subReason: AbsentSubReason?, attended: Boolean, paid: Boolean, comment: String?): EventOutcome? {
    requireTrue(attended && reason != null) { "An absent reason was supplied for a valid attendance" }
    requireTrue(!attended && reason == null) { "An absent reason was not supplied" }

    if (attended && paid) return EventOutcome(eventOutcome = "ATT", performance = "STANDARD", outcomeComment = comment)

    requireTrue(paid && !paidReasons.contains(reason)) { "$reason is not a valid paid reason" }
    requireTrue(!paid && !unpaidReasons.contains(reason)) { "$reason is not a valid unpaid reason" }

    // TODO: once DPS front end developed add in mandatory sub reason based on AbsentReason.absentSubReasonTriggers
    subReason?.run {
      requireTrue(paid && subReason == AbsentSubReason.Behaviour) { "$subReason is not a valid paid sub reason" }
    }

    return EventOutcome(
      eventOutcome = reason!!.eventOutcome,
      outcomeComment = comment?.let {
        // comment is only 240 chars in nomis, so need to truncate if we've prefixed the sub reason
        subReason?.let { "${subReason.label}. $comment" }?.take(240) ?: comment
      },
    )
  }

  private fun requireTrue(value: Boolean, lazyMessage: () -> Any) {
    require(!value, lazyMessage)
  }
}
